plugins/modules/ipadelegation.py

```diff
@@ -0,0 +1,340 @@
+
+
+def main():
+    ansible_module = AnsibleModule(
+        argument_spec=dict(
+            # general
+            ipaadmin_principal=dict(type="str", default="admin"),
+            ipaadmin_password=dict(type="str", required=False, no_log=True),
+
+            name=dict(type="list", aliases=["aciname"], default=None,
+                      required=True),
+            # present
+            permission=dict(required=False, type='list',
+                            aliases=["permissions"], default=None),
+            attribute=dict(required=False, type='list', aliases=["attrs"],
+                           default=None),
+            membergroup=dict(type="str", aliases=["memberof"], default=None),
+            group=dict(type="str", default=None),
+            action=dict(type="str", default="delegation",
+                        choices=["member", "delegation"]),
+            # state
+            state=dict(type="str", default="present",
+                       choices=["present", "absent"]),
+        ),
+        supports_check_mode=True,
+    )
+
+    ansible_module._ansible_debug = True
+
+    # Get parameters
+
+    # general
+    ipaadmin_principal = module_params_get(ansible_module,
+                                           "ipaadmin_principal")
+    ipaadmin_password = module_params_get(ansible_module, "ipaadmin_password")
+    names = module_params_get(ansible_module, "name")
+
+    # present
+    permission = module_params_get(ansible_module, "permission")
+    attribute = module_params_get(ansible_module, "attribute")
+    membergroup = module_params_get(ansible_module, "membergroup")
+    group = module_params_get(ansible_module, "group")
+    action = module_params_get(ansible_module, "action")
+    # state
+    state = module_params_get(ansible_module, "state")
+
+    # Check parameters
+
+    if state == "present":
+        if len(names) != 1:
+            ansible_module.fail_json(
+                msg="Only one delegtion be added at a time.")
+        if action == "member":
+            invalid = ["permission", "membergroup", "group"]
+            for x in invalid:
+                if vars()[x] is not None:
+                    ansible_module.fail_json(
+                        msg="Argument '%s' can not be used with action "
+                        "'%s'" % (x, action))
+
+    if state == "absent":
+        if len(names) < 1:
+            ansible_module.fail_json(msg="No name given.")
+        invalid = ["permission", "membergroup", "group"]
+        if action == "delegation":
+            invalid.append("attribute")
+        for x in invalid:
+            if vars()[x] is not None:
+                ansible_module.fail_json(
+                    msg="Argument '%s' can not be used with state '%s'" %
+                    (x, state))
+
+    if permission is not None:
+        for perm in permission:
+            if perm not in ["read", "write"]:
+                ansible_module.fail_json(msg="Invalid permission '%s'" % perm)
```

Comment:
Suggestion: to warn about all invalid permissions, this could be written as:
```python
perm = [p for p in permission if p not in ("read", "write")]
if perm:
    ansible_module.fail_json(msg="Invalid permission '%s'" % perm)
```

Reply:
Incorporated.